File: backend/app/ai/services/generation_service.py
"""
ApeAI — Generation Service

Orchestrates Gemini AI to generate structured product documents.
Enforces JSON output for all stages.
"""

import logging
import json
from typing import Dict, Any, List

from backend.app.ai.gemini_client import gemini_client
from backend.app.ai.prompts import (
    CLUSTER_SUMMARY_PROMPT,
    BRD_PROMPT,
    PRD_PROMPT,
    STORY_PROMPT,
    TASK_PROMPT
)
from backend.app.db.supabase_client import get_supabase_client
from backend.app.services.cluster_service import get_cluster, update_cluster
from backend.app.services.document_service import create_document

logger = logging.getLogger(__name__)

async def summarize_cluster(cluster_id: str) -> Dict[str, Any]:
    """Analyze all feedback in a cluster and generate a summary."""
    cluster = await get_cluster(cluster_id, include_feedback=True)
    feedback_items = cluster.get("feedback_items", [])
    
    if not feedback_items:
        raise ValueError("No feedback items found in cluster")

    feedback_text = "\n".join([
        f"- {f['content']} (Source: {f['source']})" 
        for f in feedback_items
    ])
    
    prompt = CLUSTER_SUMMARY_PROMPT.format(feedback_text=feedback_text)
    
    # Use Flash for summarization
    model = gemini_client.get_model(model_type="flash", json_mode=True)
    response = await model.generate_content_async(prompt)
    
    data = json.loads(response.text)
    
    # Update cluster with AI-generated title and summary
    await update_cluster(
        cluster_id,
        title=data["title"],
        summary=data["summary"],
        status="clustered",
        confidence_score=95.0 # Example score
    )
    
    return data

async def generate_brd(cluster_id: str) -> Dict[str, Any]:
    """Generate a BRD from a cluster summary."""
    cluster = await get_cluster(cluster_id, include_feedback=False)
    summary = cluster.get("summary")
    
    if not summary:
        raise ValueError("Cluster must be summarized before generating BRD")

    prompt = BRD_PROMPT.format(cluster_summary=summary)
    
    # Use Pro for BRD reasoning
    model = gemini_client.get_model(model_type="pro", json_mode=True)
    response = await model.generate_content_async(prompt)
    
    data = json.loads(response.text)
    
    # Save as a document
    doc = await create_document(
        cluster_id=cluster_id,
        doc_type="brd",
        title=data["title"],
        content=data
    )
    
    # Update cluster status
    await update_cluster(cluster_id, status="brd_generated")
    
    return doc

async def generate_prd(cluster_id: str, brd_id: str) -> Dict[str, Any]:
    """Generate a PRD based on an existing BRD."""
    from backend.app.services.document_service import get_document
    brd = await get_document(brd_id)
    
    prompt = PRD_PROMPT.format(brd_content=json.dumps(brd["content"]))
    
    model = gemini_client.get_model(model_type="pro", json_mode=True)
    response = await model.generate_content_async(prompt)
    
    data = json.loads(response.text)
    
    doc = await create_document(
        cluster_id=cluster_id,
        doc_type="prd",
        title=data["title"],
        content=data,
        parent_id=brd_id
    )
    
    await update_cluster(cluster_id, status="prd_generated")
    return doc

async def generate_stories(cluster_id: str, prd_id: str) -> List[Dict[str, Any]]:
    """Generate user stories from a PRD."""
    from backend.app.services.document_service import get_document
    prd = await get_document(prd_id)
    
    prompt = STORY_PROMPT.format(prd_content=json.dumps(prd["content"]))
    
    model = gemini_client.get_model(model_type="pro", json_mode=True)
    response = await model.generate_content_async(prompt)
    
    stories_data = json.loads(response.text)
    
    created_stories = []
    for story in stories_data:
        doc = await create_document(
            cluster_id=cluster_id,
            doc_type="story",
            title=story["title"],
            content=story,
            parent_id=prd_id
        )
        created_stories.append(doc)
    
    await update_cluster(cluster_id, status="stories_generated")
    return created_stories

async def generate_tasks(cluster_id: str, story_id: str) -> Dict[str, Any]:
    """Generate technical tasks for a specific story."""
    from backend.app.services.document_service import get_document
    story = await get_document(story_id)
    
    prompt = TASK_PROMPT.format(story_content=json.dumps(story["content"]))
    
    model = gemini_client.get_model(model_type="flash", json_mode=True)
    response = await model.generate_content_async(prompt)
    
    data = json.loads(response.text)
    
    doc = await create_document(
        cluster_id=cluster_id,
        doc_type="task",
        title=f"Technical Tasks: {story['title']}",
        content=data,
        parent_id=story_id
    )
    
    # Note: We only update cluster status if this was the last story or similar logic
    # but for now we'll just return it.
    return doc

File: backend-node/src/ai/generationService.ts
/**
 * ApeAI — Generation Service
 *
 * Orchestrates Gemini AI to generate structured product documents.
 * Mirrors Python ai/services/generation_service.py.
 */

import {
  CLUSTER_SUMMARY_PROMPT,
  BRD_PROMPT,
  PRD_PROMPT,
  STORY_PROMPT,
  TASK_PROMPT,
  formatPrompt,
} from './prompts';
import { geminiClient } from './geminiClient';
import { getCluster, updateCluster } from '../services/clusterService';
import { createDocument, getDocument } from '../services/documentService';

type DbRow = Record<string, unknown>;

export async function summarizeCluster(clusterId: string): Promise<DbRow> {
  const cluster = await getCluster(clusterId, true);
  const feedbackItems = (cluster['feedback_items'] ?? []) as DbRow[];

  if (feedbackItems.length === 0) throw new Error('No feedback items found in cluster');

  const feedbackText = feedbackItems
    .map((f) => `- ${f['content']} (Source: ${f['source']})`)
    .join('\n');

  const prompt = formatPrompt(CLUSTER_SUMMARY_PROMPT, { feedback_text: feedbackText });

  const model = geminiClient.getModel('flash', true);
  const response = await model.generateContent(prompt);
  const data = JSON.parse(response.response.text()) as DbRow;

  await updateCluster(clusterId, {
    title: data['title'] as string,
    summary: data['summary'] as string,
    status: 'clustered',
    confidenceScore: 95.0,
  });

  return data;
}

export async function generateBrd(clusterId: string): Promise<DbRow> {
  const cluster = await getCluster(clusterId, false);
  const summary = cluster['summary'] as string | undefined;

  if (!summary) throw new Error('Cluster must be summarized before generating BRD');

  const prompt = formatPrompt(BRD_PROMPT, { cluster_summary: summary });

  const model = geminiClient.getModel('pro', true);
  const response = await model.generateContent(prompt);
  const data = JSON.parse(response.response.text()) as DbRow;

  const doc = await createDocument({
    clusterId,
    docType: 'brd',
    title: data['title'] as string,
    content: data,
  });

  await updateCluster(clusterId, { status: 'brd_generated' });
  return doc;
}

export async function generatePrd(clusterId: string, brdId: string): Promise<DbRow> {
  const brd = await getDocument(brdId);

  const prompt = formatPrompt(PRD_PROMPT, {
    brd_content: JSON.stringify(brd['content']),
  });

  const model = geminiClient.getModel('pro', true);
  const response = await model.generateContent(prompt);
  const data = JSON.parse(response.response.text()) as DbRow;

  const doc = await createDocument({
    clusterId,
    docType: 'prd',
    title: data['title'] as string,
    content: data,
    parentId: brdId,
  });

  await updateCluster(clusterId, { status: 'prd_generated' });
  return doc;
}

export async function generateStories(clusterId: string, prdId: string): Promise<DbRow[]> {
  const prd = await getDocument(prdId);

  const prompt = formatPrompt(STORY_PROMPT, {
    prd_content: JSON.stringify(prd['content']),
  });

  const model = geminiClient.getModel('pro', true);
  const response = await model.generateContent(prompt);
  const storiesData = JSON.parse(response.response.text()) as DbRow[];

  const createdStories: DbRow[] = [];
  for (const story of storiesData) {
    const doc = await createDocument({
      clusterId,
      docType: 'story',
      title: story['title'] as string,
      content: story,
      parentId: prdId,
    });
    createdStories.push(doc);
  }

  await updateCluster(clusterId, { status: 'stories_generated' });
  return createdStories;
}

export async function generateTasks(clusterId: string, storyId: string): Promise<DbRow> {
  const story = await getDocument(storyId);

  const prompt = formatPrompt(TASK_PROMPT, {
    story_content: JSON.stringify(story['content']),
  });

  const model = geminiClient.getModel('flash', true);
  const response = await model.generateContent(prompt);
  const data = JSON.parse(response.response.text()) as DbRow;

  const doc = await createDocument({
    clusterId,
    docType: 'task',
    title: `Technical Tasks: ${story['title']}`,
    content: data,
    parentId: storyId,
  });

  return doc;
}
